# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Naiema Elsaadi

My GitHub Repository: [GitHub Repository](https://github.com/NaiemaElsaadi/Json-sentiment) 

Welcome to my Python Notebooks, Basics, and Data Structures repository! I'm excited to explore various concepts in Python programming, data structures, JSON, and applied NLP. Feel free to browse through my notebooks and projects, and don't hesitate to reach out if you have any questions or suggestions.



In [1]:

## My name is Naiema
import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list


All prereqs installed.
Package                       Version
----------------------------- -----------
annotated-types               0.6.0
anyio                         3.6.1
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
astroid                       2.11.7
asttokens                     2.0.8
attrs                         22.1.0
Babel                         2.10.3
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
beautifulsoup4                4.11.1
bleach                        5.0.1
blis                          0.7.11
brotlipy                      0.7.0
catalogue                     2.0.10
certifi                       2023.7.22
cffi                          1.15.1
charset-normalizer            2.0.4
click                         8.1.3
cloudpathlib                  0.16.0
colorama                      0.4.6
conda                         4.14.0
conda-package-handling        1.8.1
confection                    0.1.4
cryptography      

<b>Q1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [2]:
## My name is Naiema
import requests
import json

# Make API request to get song lyrics
result = json.loads(requests.get('https://api.lyrics.ovh/v1/Coldplay/Adventure of a Lifetime').text)

# Write resulting JSON to a file
with open('lyrics_song.json', 'w') as file:
    json.dump(result, file)

print("JSON data written to 'lyrics_song.json' successfully.")

JSON data written to 'lyrics_song.json' successfully.


<b>Q2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [3]:
## My name is Naiema
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

# Add the SpacyTextBlob component
nlp.add_pipe('spacytextblob')

# Read in the contents of the JSON file
with open('lyrics_song.json', 'r') as file:
    lyrics_song = json.load(file)

# Extract the lyrics from the dictionary
lyrics = lyrics_song['lyrics']

# Process the lyrics using spaCy
doc = nlp(lyrics)

# Perform sentiment analysis
sentiment_score = doc._.polarity

# Print the lyrics of the song
print("Lyrics of the song:")
print(lyrics)
print()

# Print the polarity score of the sentiment analysis
print("Polarity score of the sentiment analysis:", sentiment_score)

# Answer to the question: 
# The polarity score of approximately 0.226 suggests that the lyrics have a more positive connotation.
# Since the polarity score falls within the range [0, 1], it indicates positivity,
# it indicates that the overall sentiment expressed in the lyrics is positive.


Lyrics of the song:
Paroles de la chanson Adventure Of A Lifetime par Coldplay
Turn your magic on
To me she'd say
Everything you want's a dream away
We are legends
Every day
That's what she told him!

Turn your magic on
To me she'd say
Everything you want's a dream away
Under this pressure, under this weight
We are diamonds

I feel my heart beating
I feel my heart beneath my skin
I feel my heart beating


Ohhh, you make me feel
Like I'm alive again
Alive again
Ohhh, you make me feel
Like I'm alive again

Said I can't go on, not in this way
I'm a dream, I die by light of day
Gonna hold up half the sky and say
Ohhh, we are omen

I feel my heart beating
I feel my heart beneath my skin
Ohhh, I can feel my heart beating
Cause you make me feel
Like I'm alive again
Alive again...

Ohhh, you make me feel
Like I'm alive again

Turn your magic on, to me she'd say
Everything you want's a dream away
Under this pressure, under this weight

We are diamonds taking shape!
We are diamonds taking shape!

<b>Q3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [4]:

## My name is Naiema
import requests
import json

def get_song_lyrics(artist, song, filename):
    # Construct the API URL
    api_url = f"https://api.lyrics.ovh/v1/{artist}/{song}"

    # Make the API request
    response = requests.get(api_url)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        data = response.json()

        # Extract the lyrics
        lyrics = data.get('lyrics')

        # Write the lyrics to the specified file
        with open(filename, 'w') as file:
            file.write(lyrics)

        print(f"Lyrics for '{song}' by {artist} written to '{filename}' successfully.")
    else:
        print(f"Failed to retrieve lyrics for '{song}' by {artist} from the API.")

# Test the function with four songs and store them in different files
get_song_lyrics("Coldplay", "Adventure of a Lifetime", "adventure_of_a_lifetime_lyrics.json")
get_song_lyrics("Coldplay", "Adventure of a Lifetime", "adventure_of_a_lifetime_lyrics.txt")
get_song_lyrics("Ed Sheeran", "Shape of You", "shape_of_you_lyrics.json")
get_song_lyrics("Ed Sheeran", "Shape of You", "shape_of_you_lyrics.txt")
get_song_lyrics("Adele", "Hello", "hello_lyrics.json")
get_song_lyrics("Adele", "Hello", "hello_lyrics.txt")
get_song_lyrics("Queen", "Bohemian Rhapsody", "bohemian_rhapsody_lyrics.json")
get_song_lyrics("Queen", "Bohemian Rhapsody", "bohemian_rhapsody_lyrics.txt")


Lyrics for 'Adventure of a Lifetime' by Coldplay written to 'adventure_of_a_lifetime_lyrics.json' successfully.
Lyrics for 'Adventure of a Lifetime' by Coldplay written to 'adventure_of_a_lifetime_lyrics.txt' successfully.
Lyrics for 'Shape of You' by Ed Sheeran written to 'shape_of_you_lyrics.json' successfully.
Lyrics for 'Shape of You' by Ed Sheeran written to 'shape_of_you_lyrics.txt' successfully.
Lyrics for 'Hello' by Adele written to 'hello_lyrics.json' successfully.
Lyrics for 'Hello' by Adele written to 'hello_lyrics.txt' successfully.
Lyrics for 'Bohemian Rhapsody' by Queen written to 'bohemian_rhapsody_lyrics.json' successfully.
Lyrics for 'Bohemian Rhapsody' by Queen written to 'bohemian_rhapsody_lyrics.txt' successfully.


<b>Q4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [5]:
## My name is Naiema
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

def get_sentiment_score(filename):
    # Load the spaCy model
    nlp = spacy.load('en_core_web_sm')

    # Add the SpacyTextBlob component
    nlp.add_pipe('spacytextblob')

    # Read the lyrics from the file
    with open(filename, 'r') as file:
        lyrics = file.read()

    # Process the lyrics using spaCy
    doc = nlp(lyrics)

    # Perform sentiment analysis
    sentiment_score = doc._.polarity

    return sentiment_score

# Test the function with the filenames of the songs
files = ["adventure_of_a_lifetime_lyrics.txt", "shape_of_you_lyrics.txt", "hello_lyrics.txt", "bohemian_rhapsody_lyrics.txt"]
for file in files:
    polarity_score = get_sentiment_score(file)
    print(f"Polarity score for '{file}': {polarity_score}")


Polarity score for 'adventure_of_a_lifetime_lyrics.txt': 0.22585784313725493
Polarity score for 'shape_of_you_lyrics.txt': 0.32200413223140495
Polarity score for 'hello_lyrics.txt': -0.15366666666666665
Polarity score for 'bohemian_rhapsody_lyrics.txt': -0.0029495701058200783



<BR><B> The reported polarity scores are as follows:
<BR>`'adventure_of_a_lifetime_lyrics.txt': 0.226`
<br>`'shape_of_you_lyrics.txt': 0.322`
<br>`'hello_lyrics.txt': -0.154`
<br>`'bohemian_rhapsody_lyrics.txt': -0.003`
<br><b>The reported polarity scores generally match the understanding of the song's lyrics:
<b>"Adventure of a Lifetime" by Coldplay and "Shape of You" by Ed Sheeran have upbeat and positive themes, reflected in their positive polarity scores.
"Hello" by Adele tends to be more introspective and melancholic, resulting in a slightly negative polarity score.
"Bohemian Rhapsody" by Queen is known for its complex and ambiguous lyrics, which might explain the polarity score close to neutral.
<b>Overall, the reported polarity scores align with the expected sentiments, understanding, and themes of the above songs.














In [6]:
!jupyter nbconvert --to html requests-json-nlp.ipynb

[NbConvertApp] Converting notebook requests-json-nlp.ipynb to html
[NbConvertApp] Writing 608063 bytes to requests-json-nlp.html
